violation emails/GCS uploads now use the CSV format

arbeit opened the pull request:
Fixes issue #1048, the changes are fully tested; emails and GCS uploads work correctly.

Please ignore any commits prior to Mar 21, 2018 -- these have been merged to `2.0-dev` already (with PR #1260). The first relevant commit is 7d8bcdf030430ee1eca5a369fe3925cea0f1123c.

In any case, please take a look and let me know what you think.

blueandgold commented on google/cloud/forseti/notifier/notifiers/email_violations.py:
```diff
@@ -14,12 +14,16 @@
 
+
+import shutil
+
```
nit: remove this extra line

blueandgold commented on google/cloud/forseti/notifier/notifiers/email_violations.py:
```diff
@@ -76,8 +80,14 @@ def _write_temp_attachment(self):
+            output_csv_name = csv_file.name
+            LOGGER.info('CSV filename: %s', output_csv_name)
+
+            shutil.copyfile(output_csv_name, output_file_path)
```
Just to double-check, is the csv file being cleaned up after everything is done?

arbeit replied:
Yes, it uses a `NamedTemporaryFile` that is deleted upon context manager exit.

blueandgold commented on tests/notifier/notifiers/test_data_attachment.csv:
```diff
@@ -0,0 +1,3 @@
+resource_id,resource_type,full_name,rule_index,rule_name,violation_type,violation_data
```
nit: We have the convention of putting test data like this, into a `test_data` directory.  So, can you please create `/tests/notifier/notifiers/test_data` package, and move this there?  You will have create an `__init__.py` file too.  And then call this file `expected_attachment.csv`?

arbeit replied:
Done.

blueandgold commented on tests/notifier/notifiers/email_violations_test.py:
```diff
@@ -0,0 +1,116 @@
+        file_name = self.evp._write_temp_attachment()
+        self.assertTrue(
+            filecmp.cmp('/tmp/%s' % file_name,
+                        self.test_data_path, shallow=False))
```
When we are here, it's hard to tell what `test_data_path` is.  It would be better to call this `self.expected_attachment_path`, for example.  Then, it's super clear, what we are comparing.

arbeit replied:
Done.

arbeit commented:
Henry, I had to rebase the branch in question and open a new PR (PR #1288) -- the latter addresses all your review comments and also fixes the `email_violations` notifier that was leaking temp files indeed.
